File: app/frontend/src/domain/repositories/ReservationRepository.ts
/**
 * Repository interface for reservations
 * Defines the contract for reservation data operations
 */
import type { Reservation } from '../../types';

export interface ReservationFilters {
  from?: string;
  to?: string;
  user_id?: string;
  space_id?: string;
  status?: 'active' | 'cancelled';
}

export interface ReservationRepository {
  findAll(filters?: ReservationFilters): Promise<Reservation[]>;
  findById(id: string): Promise<Reservation>;
  create(reservation: CreateReservationInput): Promise<Reservation>;
  update(id: string, reservation: UpdateReservationInput): Promise<Reservation>;
  delete(id: string): Promise<void>;
}

export interface CreateReservationInput {
  space_id: string;
  user_id: string;
  user_name?: string;
  date: string;
  start_time?: string;
  end_time?: string;
  notes?: string;
}

export interface UpdateReservationInput {
  user_name?: string;
  date?: string;
  start_time?: string;
  end_time?: string;
  status?: 'active' | 'cancelled';
  notes?: string;
}

File: app/backend/internal/domain/repositories/reservation_repository.go
package repositories

import (
	"time"

	"github.com/google/uuid"
	"office-reservations/internal/domain/entities"
)

// ReservationRepository defines the interface for reservation data operations
type ReservationRepository interface {
	// FindByID finds a reservation by its ID
	FindByID(id uuid.UUID) (*entities.Reservation, error)
	
	// FindAll retrieves all reservations with optional filters
	FindAll(filters ReservationFilters) ([]*entities.Reservation, error)
	
	// Create creates a new reservation
	Create(reservation *entities.Reservation) error
	
	// Update updates an existing reservation
	Update(reservation *entities.Reservation) error
	
	// Delete deletes a reservation (soft delete by setting status to cancelled)
	Delete(id uuid.UUID) error
	
	// DeleteBySpaceAndTime deletes reservations for a specific space, date, and time
	DeleteBySpaceAndTime(spaceID uuid.UUID, date time.Time, startTime *string) error
	
	// DeleteBySpaceIDsAndTime deletes reservations for multiple spaces with same date and time
	DeleteBySpaceIDsAndTime(spaceIDs []uuid.UUID, date time.Time, startTime *string) error
	
	// FindBySpaceAndDate finds reservations for a specific space and date
	FindBySpaceAndDate(spaceID uuid.UUID, date time.Time) ([]*entities.Reservation, error)
	
	// FindBySpaceIDsAndDate finds reservations for multiple spaces and date
	FindBySpaceIDsAndDate(spaceIDs []uuid.UUID, date time.Time) ([]*entities.Reservation, error)
	
	// FindActiveBySpaceAndTime finds active reservations for a space, date, and time
	FindActiveBySpaceAndTime(spaceID uuid.UUID, date time.Time, startTime *string) (*entities.Reservation, error)
}

// ReservationFilters contains optional filters for querying reservations
type ReservationFilters struct {
	From    *time.Time
	To      *time.Time
	UserID  *string
	SpaceID *uuid.UUID
	Status  *entities.ReservationStatus
}

